File: app/src/main/java/com/youtube/sorcjc/redemnorte/PanelActivity.java
package com.youtube.sorcjc.redemnorte;

import android.support.design.widget.FloatingActionButton;
import android.support.v7.app.AppCompatActivity;
import android.os.Bundle;
import android.support.v7.widget.LinearLayoutManager;
import android.support.v7.widget.RecyclerView;
import android.support.v7.widget.Toolbar;

import java.util.ArrayList;

public class PanelActivity extends AppCompatActivity {

    private HeaderAdapter headerAdapter;

    @Override
    protected void onCreate(Bundle savedInstanceState) {
        super.onCreate(savedInstanceState);
        setContentView(R.layout.activity_panel);

        RecyclerView recyclerView = (RecyclerView) findViewById(R.id.recyclerView);

        LinearLayoutManager linearLayoutManager = new LinearLayoutManager(this);
        recyclerView.setLayoutManager(linearLayoutManager);

        ArrayList<Header> myDataset = new ArrayList<>();
        myDataset.add(new Header("100001", "Juan Ramos", "27 de Octubre del 2016"));
        myDataset.add(new Header("100002", "Juan Ramos", "28 de Octubre del 2016"));
        myDataset.add(new Header("100003", "Juan Ramos", "29 de Octubre del 2016"));

        myDataset.add(new Header("100004", "Juan Ramos", "27 de Noviembre del 2016"));
        myDataset.add(new Header("100005", "Juan Ramos", "28 de Noviembre del 2016"));
        myDataset.add(new Header("100006", "Juan Ramos", "29 de Noviembre del 2016"));

        myDataset.add(new Header("100007", "Juan Ramos", "27 de Diciembre del 2016"));
        myDataset.add(new Header("100008", "Juan Ramos", "28 de Diciembre del 2016"));
        myDataset.add(new Header("100009", "Juan Ramos", "29 de Diciembre del 2016"));

        headerAdapter = new HeaderAdapter(myDataset);
        recyclerView.setAdapter(headerAdapter);

        Toolbar toolbar = (Toolbar) findViewById(R.id.toolbar);
        setSupportActionBar(toolbar);

        final FloatingActionButton fab = (FloatingActionButton) findViewById(R.id.fab);

        recyclerView.addOnScrollListener(new RecyclerView.OnScrollListener()
        {
            @Override
            public void onScrolled(RecyclerView recyclerView, int dx, int dy)
            {
                if (dy > 0 || dy < 0 && fab.isShown())
                {
                    fab.hide();
                }
            }

            @Override
            public void onScrollStateChanged(RecyclerView recyclerView, int newState)
            {
                if (newState == RecyclerView.SCROLL_STATE_IDLE)
                {
                    fab.show();
                }

                super.onScrollStateChanged(recyclerView, newState);
            }
        });
    }
}

File: app/src/main/java/com/youtube/sorcjc/redemnorte/ui/activity/PanelActivity.kt
package com.youtube.sorcjc.redemnorte.ui.activity

import android.Manifest
import android.content.Intent
import android.location.Location
import android.location.LocationListener
import android.location.LocationManager
import android.os.Bundle
import android.util.Log
import android.view.Menu
import android.view.View
import androidx.appcompat.app.AppCompatActivity
import androidx.appcompat.widget.Toolbar
import androidx.fragment.app.FragmentTransaction
import androidx.recyclerview.widget.LinearLayoutManager
import androidx.recyclerview.widget.RecyclerView
import com.google.android.material.snackbar.Snackbar
import com.youtube.sorcjc.redemnorte.R
import com.youtube.sorcjc.redemnorte.io.MyApiAdapter
import com.youtube.sorcjc.redemnorte.model.Sheet
import com.youtube.sorcjc.redemnorte.ui.adapter.HeaderAdapter
import com.youtube.sorcjc.redemnorte.ui.fragment.HeaderDialog
import com.youtube.sorcjc.redemnorte.util.PreferenceHelper
import com.youtube.sorcjc.redemnorte.util.PreferenceHelper.get
import com.youtube.sorcjc.redemnorte.util.checkAndRequestPermission
import com.youtube.sorcjc.redemnorte.util.snack
import com.youtube.sorcjc.redemnorte.util.toast
import kotlinx.android.synthetic.main.activity_panel.*
import retrofit2.Call
import retrofit2.Callback
import retrofit2.Response
import java.util.*

class PanelActivity : AppCompatActivity(), View.OnClickListener, Callback<ArrayList<Sheet>> {

    private val headerAdapter by lazy {
        HeaderAdapter()
    }

    private val preferences by lazy {
        PreferenceHelper.defaultPrefs(this)
    }

    private var locationManager : LocationManager? = null

    override fun onCreate(savedInstanceState: Bundle?) {
        super.onCreate(savedInstanceState)
        setContentView(R.layout.activity_panel)

        recyclerView.layoutManager = LinearLayoutManager(this)

        loadInventorySheets()

        recyclerView.adapter = headerAdapter

        setSupportActionBar(toolbar)
        hideAndDisplayToolbarAccordingly()

        fab.setOnClickListener(this)
        trackLocationIfPermissionIsGranted()

        btnQuery.setOnClickListener(this)
    }

    private fun trackLocationIfPermissionIsGranted() {
        checkAndRequestPermission(
                getString(R.string.dialog_location_title),
                getString(R.string.dialog_location_explanation),
                Manifest.permission.ACCESS_FINE_LOCATION, REQUEST_LOCATION_PERMISSION,
                ::createLocationManagerReference
        )
    }

    private fun createLocationManagerReference() {
        // Create persistent LocationManager reference
        locationManager = getSystemService(LOCATION_SERVICE) as LocationManager?

        try {
            // Request location updates
            locationManager?.requestLocationUpdates(LocationManager.NETWORK_PROVIDER, 0L, 0f, locationListener)
        } catch(ex: SecurityException) {
            Log.d("PanelActivity", "Security Exception, no location available")
        }
    }

    private val locationListener: LocationListener = object : LocationListener {
        override fun onLocationChanged(location: Location) {
            postUserLocation(location.latitude, location.longitude)
        }

        override fun onStatusChanged(provider: String, status: Int, extras: Bundle) {}
        override fun onProviderEnabled(provider: String) {}
        override fun onProviderDisabled(provider: String) {}
    }

    private fun postUserLocation(lat: Double, lng: Double) {
        Log.d("PanelActivity", "longitude $lng, latitude $lat")

        MyApiAdapter.getApiService()
                .postUserLocation(preferences["user_id"], lat, lng)
                .enqueue(object : Callback<Void> {
                    override fun onResponse(call: Call<Void>, response: Response<Void>) {
                        if (response.isSuccessful) {
                            Log.d("PanelActivity", "userLocation sent to the server")
                        } else {
                            Log.d("PanelActivity", "responseCode ${response.code()}")
                        }
                    }

                    override fun onFailure(call: Call<Void>, t: Throwable) {
                        toast(t.localizedMessage ?: "")
                    }
                })
    }

    private fun hideAndDisplayToolbarAccordingly() {
        recyclerView.addOnScrollListener(object : RecyclerView.OnScrollListener() {
            override fun onScrolled(recyclerView: RecyclerView, dx: Int, dy: Int) {
                if (dy > 0 || dy < 0 && fab.isShown) {
                    fab.hide()
                }
            }

            override fun onScrollStateChanged(recyclerView: RecyclerView, newState: Int) {
                if (newState == RecyclerView.SCROLL_STATE_IDLE) {
                    fab.show()
                }
                super.onScrollStateChanged(recyclerView, newState)
            }
        })

        toolbar.setOnMenuItemClickListener(
                Toolbar.OnMenuItemClickListener { item ->
                    val id = item.itemId
                    if (id == R.id.itemSearch) {
                        actionSearch()
                    } else if (id == R.id.itemSignature) {
                        actionSignature()
                    }
                    return@OnMenuItemClickListener true
                })
    }

    private fun actionSearch() {
        if (layoutSearchBox.visibility == View.VISIBLE)
            layoutSearchBox.visibility = View.GONE
        else
            layoutSearchBox.visibility = View.VISIBLE
    }

    private fun actionSignature() {
        val intent = Intent(this, SignatureActivity::class.java)
        startActivity(intent)
    }

    override fun onCreateOptionsMenu(menu: Menu): Boolean {
        val inflater = menuInflater
        inflater.inflate(R.menu.search_menu, menu)
        return super.onCreateOptionsMenu(menu)
    }

    fun loadInventorySheets() {
        val userId = preferences["user_id", -1]

        MyApiAdapter.getApiService()
                .getSheets(userId)
                .enqueue(this)
    }

    override fun onClick(view: View) {
        when (view.id) {
            R.id.fab -> showCreateHeaderDialog()
            R.id.btnQuery ->
                headerAdapter.filterResponsibleStartsWith(etQueryHeader.text.toString().trim())
        }
    }

    private fun showCreateHeaderDialog() {
        val fragmentManager = supportFragmentManager

        // Empty sheetId => Create a new sheet
        val newFragment = HeaderDialog.newInstance(-1)
        val transaction = fragmentManager.beginTransaction()

        transaction.setTransition(FragmentTransaction.TRANSIT_FRAGMENT_OPEN)
        transaction.add(android.R.id.content, newFragment)
                .addToBackStack(null).commit()
    }

    override fun onResponse(call: Call<ArrayList<Sheet>>, response: Response<ArrayList<Sheet>>) {
        if (response.isSuccessful) {
            response.body()?.let { sheets ->
                headerAdapter.setDataSet(sheets)
                panelRootLayout.snack(getString(R.string.sheets_count_message, sheets.size))
            }
        } else {
            toast(getString(R.string.error_format_server_response))
        }
        showRecyclerView()
    }

    override fun onFailure(call: Call<ArrayList<Sheet>>, t: Throwable) {
        toast(t.localizedMessage ?: "")
        showRecyclerView()
    }

    private fun showRecyclerView() {
        progressBarSheets.visibility = View.GONE

        recyclerView.visibility = View.VISIBLE
        fab.show()
    }

    companion object {
        private const val REQUEST_LOCATION_PERMISSION = 10100
    }
}